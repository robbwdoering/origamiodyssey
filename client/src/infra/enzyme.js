/**
 * FILENAME: enzyme.js
 *
 * DESCRIPTION: Setup for AirBnB's Enzyme package. 
 */

import Enzyme, { configure, shallow, mount, render } from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';

// Spy on the authentication hook, so we can pass whatever we want
// src: https://stackoverflow.com/questions/67509953/mocking-auth0client-in-auth0-in-jest
jest.mock('@auth0/auth0-react', () => ({
    Auth0Provider: ({ children }) => children,
    withAuthenticationRequired: (component => component),
    useAuth0: () => ({
        isLoading: false,
        user: { sub: "TEST_USER" },
        isAuthenticated: true,
        loginWithRedirect: jest.fn(),
        logout: jest.fn()
    })
}));


configure({ adapter: new Adapter() });
export { shallow, mount, render };
export default Enzyme;
