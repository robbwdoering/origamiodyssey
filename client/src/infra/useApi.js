/**
 * FILENAME: useApi.js
 *
 * DESCRIPTION: Helper hook that manages a connection to the authenticated REST API that backs the app.
 * src: https://github.com/auth0/auth0-react/blob/master/EXAMPLES.md#4-create-a-useapi-hook-for-accessing-protected-apis-with-an-access-token
 * @param path JUST THE PATH, not the full URL, of the requested resource
 * @param options object defining audience, scope, and any custom inputs
 */
import { useState } from 'react';
import { useAuth0 } from '@auth0/auth0-react';

export const useApi = (url, method, options = {}, handleFetch, data) => {
    const { getAccessTokenSilently, user, isAuthenticated } = useAuth0();
    const [state, setState] = useState({
        error: null,
        loading: false,
        data: null
    });

    const sleep = ms => {
        return new Promise(resolve => setTimeout(resolve, ms));
    }

    const perform = async (urlSuffix, overrideValues, overrideMethod) => {
        let count = 20;
        while ((!isAuthenticated) && count > 0) {
            count--;
            await sleep(1000);
        }

        try {
            const accessToken = await getAccessTokenSilently(options);

            let msg = {
                method: overrideMethod || method,
                headers: {
                    Accept: 'application/json',
                    'Content-Type': 'application/json',
                    // Authorization: `Bearer ${accessToken}`,
                    strategy: 'auth0',
                    access_token: accessToken
                }
            };

            // Append data if it was passed
            if (overrideValues) {
                msg.body = JSON.stringify(overrideValues);
            } else if (data && method !== "GET") {
                msg.body = JSON.stringify(data);
            }
            setState(Object.assign({}, state, { loading: true }))

            // Perform the fetch asynchrounously
            fetch(process.env.REACT_APP_API_IP + url + (urlSuffix || ""), msg)
            .then(response => {
                if (!response.ok) {
                    console.error('[useApi] Received a bad status from the server: ', response);
                    throw new Error(`status ${response.status}`);
                }
                return response.json();
            })
            .then((json) => {
                // This line handles all responses - created at app root to hook into redux
                if (handleFetch) {
                    handleFetch(json);
                }

                setState(Object.assign({}, state, {data: json}))

                // Turn off loading indicator
                setState(Object.assign({}, state, { loading: false }));
            })
            .catch(e => {
                console.log('call failed', e);
            });
        } catch (error) {
            console.log('useApi FAILURE from exception: ', error);
            setState({
                ...state,
                error,
                loading: false
            });
        }
    };

    return {
    ...state,
    refresh: perform
    };
};
