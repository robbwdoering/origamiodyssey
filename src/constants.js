/**
 * FILENAME: Constants.js 
 *
 * DESCRIPTION: Contains constants for use accross the app. 
 */

export const Pages = {
	Splash: "Splash",
	Lesson: "Lesson"
};

export const initAppReducerState = {
	page: Pages.Splash
};
