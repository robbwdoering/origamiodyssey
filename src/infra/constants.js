/**
 * FILENAME: Constants.js 
 *
 * DESCRIPTION: Contains constants for use accross the app. 
 */

import React, { useState, useRef, useMemo, useEffect } from 'react';
import GitHubIcon from "@material-ui/icons/GitHub";
import SchoolIcon from "@material-ui/icons/School";
import ClassIcon from "@material-ui/icons/Class";
import DashboardIcon from "@material-ui/icons/Dashboard";
import GroupAddIcon from "@material-ui/icons/GroupAdd";

// The FOLD files 
import BirdBase from "./../folds/BirdBase.json";
import BoatBase from "./../folds/BoatBase.json";
import FrogBase from "./../folds/FrogBase.json";
import SailBoat from "./../folds/SailBoat.json";

export const Pages = {
	Splash: "Splash",
	ModelSelect: "ModelSelect",
	Fold: "Fold",
	User: "User",

	// Used but undefined
	Lesson: "Lesson",
	History: "History",
	LearnDashboard: "LearnDashboard",
	TeachDashboard: "TeachDashboard"
};

export const Tags = {
	base: {
		text: "Base",
		category: "family"
	},
	bird: {
		text: "Bird",
		category: "counterpart",
	},
	boat: {
		text: "Boat",
		category: "counterpart"
	},
	flower: {
		text: "Flower",
		category: "counterpart"
	}
};

export const Folds = {
	BirdBase: {
		name: "Bird Base",
		description: "The first base many learn, this is perfect for winged creations such as birds and dragons.",
		tags: ["base"],
		author: "Traditional",
		json: BirdBase
	},
	BoatBase: {
		name: "Boat Base",
		description: "A more specific base for blocky models, such as boats and frames.",
		tags: ["base"],
		author: "Traditional",
		json: BoatBase
	},
	FrogBase: {
		name: "Frog Base",
		description: "A complex base used for models with length-wise asymmetry, like flowers and animals.",
		tags: ["base"],
		author: "Traditional",
		json: FrogBase 
	},
	SailBoat: {
		name: "Sail Boat",
		description: "An assymetrical sailboat with a built in kickstand - perfect for display.",
		tags: ["boat"],
		author: "Traditional",
		json: SailBoat 
	}
};

export const initAppReducerState = {
	layoutState: {
		hash: 0,
		page: Pages.ModelSelect,
		curFold: null,
		foldHash: 0,
		useImages: false,
		showEditor: true,
		expandHierarchy: false 
	},
	foldState: {
		hash: 0,
		selectedLevel: 0,
		stepIdx: -1, // Depends on selected level
		overrideWithEditor: false,
		usingDefaults: true
	},
	editorState: {
		hash: 0,
		edgeHighlights: [],
		vertexHighlights: [],
		faceHighlights: [],
		showEdges: true,
		showVertices: true,
		showFaces: true,
		showTriangulations: false,
		showLables: false
	},
	showNavDrawer: false

};

export const Actions = {
	SET_LAYOUT_STATE: "SET_LAYOUT_STATE",
	SET_SHOW_NAV_DRAWER: "SET_SHOW_NAV_DRAWER",
	SET_PAGE_CONFIG: "SET_PAGE_CONFIG",
	SET_FOLD_STATE: "SET_FOLD_STATE",
	SET_EDITOR_STATE: "SET_EDITOR_STATE",
};

export const initNavTree = [
	{	
		text: "What is Origami?",
		key: "context_root",
		children: [
			{
				text: "History",
				key: "history"				
			},
			{
				text: "Origami Today",
				key: "current_origami"
			},
			{
				text: "Practical Origami",
				key: "practical"
			}
		]
	},
	{	
		text: "Learn Origami",
		key: "learn_root",
		children: [
			{
				icon: <DashboardIcon />,
				text: "Dashboard",
				key: "learn_dashboard"				
			},
			{
				icon: <GroupAddIcon />,
				text: "Join Class",
				key: "join_class"
			}
		]
	},
	{	
		text: "Teach Origami",
		key: "teach_root",
		children: [
			{
				icon: <DashboardIcon />,
				text: "Lesson Plans",
				key: "lesson_plan_dashboard"				
			},
			{
				icon: <SchoolIcon />,
				text: "Read Research Paper",
				key: "google_scholar"
			},
			{
				icon: <ClassIcon />,
				text: "Origami in the Classroom",
				key: "teaching_tips"
			}
		]
	},
	{	
		icon: <GitHubIcon />,
		text: "Source Code",
		key: "github"
	},
];

export const initFoldState = {
	//place to store buffer geo vertex data
	positions: [],
	//place to store buffer geo vertex colors
	colors: [],
	indices: [],
	nodes: [], 
	faces: [], 
	edges: [], 
	creases: [], 
	vertices: [], //indexed vertices array
	fold: null,
	creaseParams: null
}

/*
export const Lines = {
    U: hingeLines,
    M: mountainLines,
    V: valleyLines,
    C: cutLines,
    F: facetLines,
    B: borderLines
};
*/
