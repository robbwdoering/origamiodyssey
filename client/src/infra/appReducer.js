/**
 * FILENAME: appReducer.js
 *
 * DESCRIPTION: Handles state update for all layout actions.
 */

import supermemo2 from 'supermemo2';
import { initAppReducerState, Actions } from './constants';

const finalInitState = JSON.parse(JSON.stringify(initAppReducerState));

export const appReducer = (state = finalInitState, action) => {
	let newState = Object.assign({}, state);

	switch (action.type) {
		case Actions.SET_SHOW_NAV_DRAWER:
			// If passed a val set to that, otherwise toggle
			newState.showNavDrawer = action.payload !== undefined ? action.payload : !newState.showNavDrawer;
			break;
		case Actions.SET_LAYOUT_STATE:
			console.log('[SET_LAYOUT_STATE]', action.payload);
			Object.assign(newState.layoutState, action.payload || initAppReducerState.layoutState);
			newState.layoutState.hash++;
			break;
		case Actions.SET_FOLD_STATE:
			console.log('[SET_FOLD_STATE]', action.payload);
			Object.assign(newState.foldState, action.payload || initAppReducerState.foldState);
			newState.foldState.hash++;
			break;
		case Actions.SET_EDITOR_STATE:
			console.log('[SET_EDITOR_STATE]', action.payload);
			Object.assign(newState.editorState, action.payload || initAppReducerState.editorState);
			newState.editorState.hash++;
			break;
		case Actions.SET_USER_STATE:
			console.log('[SET_USER_STATE]', action.payload);
			Object.assign(newState.userState, action.payload || initAppReducerState.userState);
			newState.userState.hash++;
			break;
		case Actions.ADD_HISTORY_ENTRY:
			if (!action.payload) {
				return newState;
			}
			newState.userState.foldHistory.push(action.payload)

			// If this model type is part of the memo algorithm, update that entry 
			const idx = newState.userState.modelList.findIndex(model => model.foldKey === action.payload.foldKey);
			if (idx !== -1) {
				console.log("Modifying entry for ", action.payload.foldKey);

				// Update the spaced learning params given this new information
				const quality = action.payload.quality;
				const lastSchedule = newState.userState.modelList[idx].schedule
				const lastFactor = newState.userState.modelList[idx].factor
				newState.userState.modelList[idx] = {
					foldKey: action.payload.foldKey,
					...supermemo2(quality, lastSchedule, lastFactor)
				}
			}
			// Else don't rememeber anything other than the foldHistory row

			newState.userState.hash++;
			console.log('[ADD_HISTORY_ENTRY]', action.payload, newState.userState);
			break;
	}

	return newState;
};
