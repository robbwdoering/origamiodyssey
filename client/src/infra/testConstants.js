/**
 * FILENAME: testConstants.js
 *
 * DESCRIPTION: Contains constants for use in the test suites. 
 */

import { initAppReducerState } from "./constants";

export const testRedux = Object.assign({}, initAppReducerState, {
    foldState: Object.assign({}, initAppReducerState.foldState, {
        maxSteps: 10
    })
});
