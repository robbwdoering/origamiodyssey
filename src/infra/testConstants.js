/**
 * FILENAME: testConstants.js
 *
 * DESCRIPTION: Contains constants for use in the test suites. 
 */

import { initAppReducerState } from "./constants";

export const testRedux = Object.assign({}, initAppReducerState, {});
