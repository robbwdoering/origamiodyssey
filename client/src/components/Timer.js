/**
 * FILENAME: Timer.js
 *
 * DESCRIPTION: This component shows a simple timer that records to redux the amount of time taken to complete a set of instructions. 
 */

// React + Redux
import React, { useState, useRef, useMemo, useEffect, createRef } from 'react';
import { connect } from 'react-redux';

import { useUpdate, useSpring, useSprings, animated, config } from 'react-spring';

import {
	ButtonGroup,
	Divider,
	Typography,
	CardMedia,
	CardActionArea,
	CardActions,
	CardContent,
	Button,
	Grid,
	Fab,
	Card,
	Chip,
	Snackbar,
	Paper,
	Tooltip
} from '@material-ui/core';
import MuiAlert from "@material-ui/lab/Alert";
import FilterList from '@material-ui/icons/FilterList';
import Clear from '@material-ui/icons/Clear';
import PlayArrow from '@material-ui/icons/PlayArrow';
import Pause from '@material-ui/icons/Pause';
import Star from '@material-ui/icons/Star';
import StarBorder from '@material-ui/icons/StarBorder';
import Done from '@material-ui/icons/Done';

import useStyles from './../style/theme';
import { Folds, Pages, Tags, TagCategories, LikertTitles } from './../infra/constants';
import { setLayoutState, setFoldState, addHistoryEntry } from './../infra/actions';
import { timerPosixToString } from './../infra/utils';

/**
 * A timer that will re-rerender every second minimum.
 */
export const Timer = props => {
	const {
		placeholderRef,
		isActive,
		isHidden,
		layoutState,
		setLayoutState,
		foldState,
		setFoldState,
		userState,
		addHistoryEntry
	} = props;
	const [curHash, setHash] = useState(0);
	const [isPlaying, setIsPlaying] = useState(false);
	const [startPosix, setStartPosix] = useState(-1);
	const [workerId, setWorkerId] = useState(-1);
	const [hasInitialized, setHasInitialized] = useState(false);
	const [hasFinished, setHasFinished] = useState(false);
	const [showSnackbar, setShowSnackbar] = useState(false);
	const [showLikertAssess, setShowLikertAssess] = useState(false);
	const [lastLikert, setLastLikert] = useState(-1);
	const classes = useStyles();
	const style = useRef({});

	const triggerRerender = () => {
		setHash(cur => cur + 1);
	};

	const toggleTimer = () => {
		if (isPlaying && workerId !== -1) {
			setFoldState({
				lastRecordedTimer: foldState.lastRecordedTimer + (Date.now() - startPosix)
			});

			clearInterval(workerId);
			setWorkerId(-1);
			setStartPosix(-1);
		} else if (!isPlaying && workerId === -1){
			setWorkerId(setInterval(triggerRerender, 1000))
			setStartPosix(Date.now());
		}

		setIsPlaying(cur => !cur);
	};

	/**
	 * Reset the state that backs the value of the timer. Does not affect the running state;
	 * a running timer that is reset will immediately start counting from 0. 
	 */
	const resetTimer = () => {
		// Erase any record of the previous run - it is not saved anywhere
		setFoldState({ lastRecordedTimer: 0 });

		// Start running immediately if it was already running
		setStartPosix(isPlaying ? Date.now()  : -1);

		// Don't update the worker - if it's running, let it continue running
	};

	const resetAllState = () => {
		setIsPlaying(false);
		setHasInitialized(false);
		setHasFinished(false);
		setStartPosix(-1);

		setFoldState({lastRecordedTimer: 0});
	};

	const handleRecordQuality = () => {
		setShowLikertAssess(true);
	};

	const handleModelSelect = () => {
		closeSnackbar();
		setLayoutState({
			page: Pages.ModelSelect,
			curFold: null
		});

		// Reset fold state
		setFoldState(null);
	};

	const handleFoldAnother = () => {
		closeSnackbar();
		resetAllState();

		// Reset fold state
		setFoldState({
			stepIdx: -1	,
			repeatRoot: -1,
			repeatRange: null
		});
	}

	const handleSubmitLikert = () => {
		setShowLikertAssess(false);
	};

	const genLikertScale = () => {
		let ret = [];
		for (let i = 0; i < 5; i++) {
			const isActive = i <= lastLikert;
			ret.push(
				<Tooltip
					title={LikertTitles[i]}
					placement="bottom-start"
					classes={{ popper: classes.hier_node_tooltip }}
					key={`likert-tt-${i}`}
				>
					<div key={i} className={classes.likert_icon_container} onClick={() => setLastLikert(cur => cur === i ? -1 : i)}>
						{isActive ? <Star className={classes.likert_icon}/> : <StarBorder className={classes.likert_icon}/>}
					</div>
				</Tooltip>
			);
		}

		return (
			<ButtonGroup variant="text">
				{ret}
				<Button onClick={handleSubmitLikert} disabled={lastLikert === -1}>
					Submit	
				</Button>
			</ButtonGroup>
		);
	};

	const closeSnackbar = () => {
		// console.log("[closeSnackbar]", hasFinished, showSnackbar, lastLikert, layoutState.curFold);
		if (!showSnackbar) {
			return;
		}

		setShowSnackbar(false);

		// If details weren't already sent, send them now
		if (!hasFinished) {
			setHasFinished(true);	

			let newEntry = {
				foldKey: layoutState.curFold,
				time: Date.now(),
				quality: lastLikert !== -1 ? lastLikert : 3,
				timer: foldState.lastRecordedTimer
			};

			addHistoryEntry(newEntry);

			setFoldState({ lastRecordedTimer: 0 });
			setLastLikert(-1);
		}
	};

	// LIFECYCLE

	// Dynamically calculate the target size of the card
	style.current = {
		height: isActive ? '400px' : '180px',
		width: isActive ? '400px' : '200px',
		display: isHidden ? 'none' : undefined
	};

	if (placeholderRef && placeholderRef.current) {
		style.current.top = placeholderRef.current.offsetTop;
		style.current.left = placeholderRef.current.offsetLeft;
	}

	// Every time a new model is selected, the timer starts from scratch
	useEffect(resetAllState, [layoutState.curFold]);

	// Every time a step is advanced, check for completion
	useEffect(() => {
		// If this is a change to the last step, 
		if (foldState.stepIdx >= foldState.maxSteps - 1) {
			if (isPlaying) {
				toggleTimer();
			}

			setShowSnackbar(true);

			// Start the timer the first time the user changes steps
			// TODO - probably good feature, but system needs redesign
		}

		// Close the snackbar on unmount if it's open, which involves saving
		return closeSnackbar;
	}, [foldState.stepIdx]);

	let timerPosix = foldState.lastRecordedTimer;
	if (isPlaying) {
		timerPosix += Date.now() - startPosix;
	}

	// console.log("[Timer]", timerPosix, foldState.lastRecordedTimer, isPlaying);

	return (
		<React.Fragment>
			<div className={classes.fold_timer_container}>
				<ButtonGroup id="oo-timer-container">
					<Button
						className={`${classes.fold_timer} ${!isPlaying ? classes.fold_timer__paused : ''}`}
						color={isPlaying ? undefined : "primary"}
						onClick={toggleTimer}
					>
						{!isPlaying && <PlayArrow/>}
						{timerPosixToString(timerPosix)}
					</Button>
					<Button className={classes.fold_timer_control} onClick={resetTimer} disabled={!isPlaying && !foldState.lastRecordedTimer}>
						<Clear />	
					</Button>
				</ButtonGroup>
			</div>
			<Snackbar className={classes.fold_timer_snackbar} open={showSnackbar} anchorOrigin={{ vertical: "top", horizontal: "center" }}>
				<Paper elevation={5}>
				{showLikertAssess ? (
					<Grid className={classes.fold_timer_grid} container>
						<Grid item xs={12}>
							<Typography variant="h4" component="h4">
								Please rate the quality of this result...	
							</Typography>
						</Grid>
						<Grid item xs={12}>
							{genLikertScale()}
						</Grid>
					</Grid>
				) : (
					<Grid className={classes.fold_timer_grid} container>
						<Grid item xs={10}>
							<Typography  variant="h4" component="h4">
								Model complete
							</Typography>
							<Button classes={{root: classes.fold_timer_snackbar_close}} onClick={closeSnackbar}>
								<Clear />
							</Button>
						</Grid>
						{userState.showLikertAssess && (
							<Grid item xs={12} md={4}>
								<Button onClick={handleRecordQuality} disabled={lastLikert !== -1}>
									{lastLikert !== -1 && (
										<Done className={classes.fold_timer_done_icon} />
									)}
									Record Quality
								</Button>
							</Grid>
						)}
						<Grid item xs={12} md={4}>
							<Button onClick={handleFoldAnother}>
								Fold Another	
							</Button>
						</Grid>
						<Grid item xs={12} md={4}>
							<Button onClick={handleModelSelect}>
								Find Another Model	
							</Button>
						</Grid>
					</Grid>
				)}
				</Paper>
			</Snackbar>
			</React.Fragment>
	);
};

export const mapStateToProps = (state, props) => {
	return {
		layoutState: state.appReducer.layoutState,
		layoutStateHash: state.appReducer.layoutState.hash,
		foldState: state.appReducer.foldState,
		foldStateHash: state.appReducer.foldState.hash,
		userState: state.appReducer.userState,
		userStateHash: state.appReducer.userState.hash
	};
};

export default connect(mapStateToProps, { setLayoutState, setFoldState, addHistoryEntry })(Timer);
