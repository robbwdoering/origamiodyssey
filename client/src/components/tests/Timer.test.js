/**
 * FILENAME: Timer.test.js
 */

// React + Enzyme
import React from "react";
import { act } from 'react-dom/test-utils';
import { mount } from "./../../infra/enzyme";

// Local
import { testRedux } from "./../../infra/testConstants";
import { Folds } from "./../../infra/constants";
import { Timer } from "./../Timer";

describe("Timer", () => {
	let comp;
	let setLayoutState, setFoldState, addHistoryEntry;

	beforeEach(() => {
		setLayoutState = jest.fn();
		setFoldState = jest.fn();
		addHistoryEntry = jest.fn();

		comp = mount(
			<Timer
				isHidden={false}
				isActive={false}
				placeholderRef={{ current: {offsetTop: 5, offsetLeft: 10} }}

				layoutState={Object.assign({}, testRedux.layoutState)}
				layoutStateHash={0}
				foldState={Object.assign({}, testRedux.foldState)}
				foldStateHash={0}
				userState={Object.assign({}, testRedux.userState)}
				userStateHash={0}
				setLayoutState={setLayoutState}
				setFoldState={setFoldState}
				addHistoryEntry={addHistoryEntry}
			/>
		);
	});

	afterEach(() => {
		comp.unmount();
	});

	it("renders without crashing", () => {
		expect(comp).toMatchSnapshot();
	});

	it("starts the timer when clicked", async () => {
		await act(async () => {
			// Start the timer
			comp.find('button.MuiButton-root').first().simulate('click')

			// Wait 2 seconds
			await new Promise((r) => setTimeout(r, 2001));

			// Stop the timer, so it sends the result to redux
			comp.find('button.MuiButton-root').first().simulate('click')
		});

		// Check the last redux change to go out (first call is with 0)
		expect(setFoldState.mock.calls.length).toBe(2);
		expect(setFoldState.mock.calls[1][0].lastRecordedTimer > testRedux.foldState.lastRecordedTimer).toBe(true);
	});

	it("displays no 'play' arrow when already playing", () => {
		// Start the timer
		act(() => {
			comp.find('button.MuiButton-root').first().simulate('click');
		});
		comp.update();

		expect(comp.exists('#oo-timer-play-icon')).toBe(false);
	});

	it("displays a play arrow when paused", () => {
		expect(comp.find('#oo-timer-play-icon').exists()).toBe(true);
	});

	it("resets the timer when clicked", async () => {
		await act(async () => {
			// Start the timer
			comp.find('button.MuiButton-root').first().simulate('click')

			// Wait 2 seconds
			await new Promise((r) => setTimeout(r, 2500));
		});

		// Check that the text of the timer advanced
		expect(comp.find('button.MuiButton-root').first().text()).toBe('00:02');
	});


	it("shows snackbar upon completion", () => {
		// Move this to the last step
		comp.setProps({ foldState: Object.assign({}, testRedux.foldState, { stepIdx: testRedux.foldState.maxSteps }) });
		comp.update();

		expect(comp.exists('#oo-timer-snackbar'));
		expect(comp.find('#oo-timer-snackbar').first().prop('open')).toBe(true);
	});

	it("sends details upon closing snackbar", () => {
		// Move this to the last step
		comp.setProps({ foldState: Object.assign({}, testRedux.foldState, { stepIdx: testRedux.foldState.maxSteps }) });
		comp.update();

		// Close the snackbar by navigating to the models page
		comp.find("#oo-timer-snackbar").last().find('button').last().simulate('click');
		comp.update();

		// Check 'model select'
		expect(setLayoutState).toHaveBeenCalled();
		expect(setFoldState).toHaveBeenCalled();
		expect(comp.find('#oo-timer-snackbar').first().prop('open')).toBe(false);
		expect(addHistoryEntry.mock.calls.length).toBe(1);
	});

	it("only sends details once upon closing snackbar", () => {
		// Move this to the last step
		comp.setProps({ foldState: Object.assign({}, testRedux.foldState, { stepIdx: testRedux.foldState.maxSteps }) });
		comp.update();

		// Close the snackbar by folding another
		comp.find("#oo-timer-snackbar").last().find('button').last().simulate('click');
		comp.update();

		// Move this to the last step again
		comp.setProps({ foldState: Object.assign({}, testRedux.foldState, { stepIdx: testRedux.foldState.maxSteps }) });
		comp.update();

		// Close the snackbar by folding another
		comp.find("#oo-timer-snackbar").last().find('button').last().simulate('click');
		comp.update();

		// Check 'fold another'
		expect(setLayoutState).toHaveBeenCalled();
		expect(setFoldState).toHaveBeenCalled();
		expect(comp.find('#oo-timer-snackbar').first().prop('open')).toBe(false);
		expect(addHistoryEntry.mock.calls.length).toBe(1); // important: NOT 2
	});

	it("generates a Likert scale without crashing", () => {
		// Move this to the last step to show snackbar
		comp.setProps({
			foldState: Object.assign({}, testRedux.foldState, { stepIdx: testRedux.foldState.maxSteps }),
			userState: Object.assign({}, testRedux.userState, { showLikertAssess: true })
		});
		comp.update();

		// Open the quality assesment snackbar
		const button = comp.find("#oo-timer-snackbar").last().find('button').at(1);
		button.simulate('click');
		comp.update();

		expect(comp.find("#oo-timer-snackbar").first()).toMatchSnapshot();
	})
});
