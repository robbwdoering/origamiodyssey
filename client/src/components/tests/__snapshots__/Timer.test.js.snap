// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timer generates a Likert scale without crashing 1`] = `ReactWrapper {}`;

exports[`Timer renders without crashing 1`] = `ReactWrapper {}`;
